File: frontend/src/app/core/services/revision-externa.service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { API } from '../constants/api.constants';
import {
  RevisionDetalle,
  RevisionMetricas,
  RevisionResumen,
} from '../models/revision-externa.model';

/**
 * Funciones del rol revisor: subir un expediente completo (PDF) para análisis
 * con IA y consultar el historial de revisiones.
 */
@Injectable({
  providedIn: 'root',
})
export class RevisionExternaService {
  private readonly http = inject(HttpClient);

  /** Sube el PDF. El backend responde 202 con el id y analiza en segundo plano. */
  crear(archivo: File): Observable<{ id: number }> {
    const form = new FormData();
    form.append('archivo', archivo);
    return this.http.post<{ id: number }>(API.REVISOR_REVISIONES, form);
  }

  /** Historial del revisor, más nueva primero. */
  listar(): Observable<RevisionResumen[]> {
    return this.http.get<RevisionResumen[]>(API.REVISOR_REVISIONES);
  }

  /**
   * Métricas agregadas del historial. Van por endpoint aparte porque los
   * problemas detectados están dentro del `resultado` de cada revisión y el
   * listado no lo trae.
   */
  metricas(): Observable<RevisionMetricas> {
    return this.http.get<RevisionMetricas>(API.REVISOR_METRICAS);
  }

  /** Detalle con el resumen. El front lo consulta hasta que estado deje de ser EN_PROGRESO. */
  obtener(id: number): Observable<RevisionDetalle> {
    return this.http.get<RevisionDetalle>(API.REVISOR_REVISION_BY_ID(id));
  }

  eliminar(id: number): Observable<void> {
    return this.http.delete<void>(API.REVISOR_REVISION_BY_ID(id));
  }
}

File: backend/src/main/java/ar/edu/utn/frc/previsar/services/RevisionExternaService.java
package ar.edu.utn.frc.previsar.services;

import ar.edu.utn.frc.previsar.dtos.response.RevisionExternaDetalleDto;
import ar.edu.utn.frc.previsar.dtos.response.RevisionExternaResumenDto;
import ar.edu.utn.frc.previsar.dtos.response.RevisionMetricasDto;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

/**
 * Análisis de expedientes completos que sube un revisor.
 *
 * Todas las operaciones exigen que el usuario autenticado tenga rol de revisor
 * y operan solo sobre SUS propias revisiones (el historial es por usuario).
 */
public interface RevisionExternaService {

    /**
     * Valida el rol de revisor y el PDF (tipo y tamaño), lo almacena y crea la
     * revisión en estado EN_PROGRESO. Devuelve el id para que el controller
     * dispare el análisis asíncrono. NO llama a Gemini (eso es async).
     */
    Long crear(MultipartFile archivo);

    /**
     * Analiza el PDF con IA en segundo plano y persiste el resultado (COMPLETADO
     * con resumen, o ERROR con motivo visible). Se invoca cross-bean desde el
     * controller para que aplique el proxy @Async.
     */
    void analizarAsync(Long revisionId);

    /** Historial del revisor: sus revisiones, más nueva primero. */
    List<RevisionExternaResumenDto> listar();

    /**
     * Métricas agregadas del historial propio. Existe como endpoint aparte
     * porque los problemas detectados viven dentro del `resultado` de cada
     * revisión, que el listado no devuelve: sin esto el front tendría que pedir
     * el detalle de cada una para poder contarlos.
     */
    RevisionMetricasDto metricas();

    /** Detalle de una revisión propia (con el resumen). 404 ante revisiones ajenas. */
    RevisionExternaDetalleDto obtener(Long id);

    /** Elimina una revisión propia y su PDF. 404 ante revisiones ajenas. */
    void eliminar(Long id);
}
